File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/client/renderer/chat/ChatRenderer.java
package pize.tests.voxelgame.client.renderer.chat;

import pize.Pize;
import pize.app.Disposable;
import pize.graphics.texture.Region;
import pize.graphics.texture.Texture;
import pize.graphics.util.batch.TextureBatch;
import pize.math.Maths;
import pize.tests.voxelgame.client.chat.Chat;
import pize.tests.voxelgame.client.chat.ChatMessage;
import pize.tests.voxelgame.client.renderer.GameRenderer;
import pize.tests.voxelgame.client.renderer.text.TextComponentBatch;
import pize.tests.voxelgame.main.text.Component;
import pize.tests.voxelgame.main.text.ComponentText;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ChatRenderer implements Disposable{
    
    private static final float MSG_LIFE_TIME_SEC = 6;
    
    private final GameRenderer gameRenderer;
    private final TextureBatch batch;
    private final TextComponentBatch textBatch;
    private final int chatX, chatY;
    private float scroll;
    private float scrollMotion;
    
    private final Map<Integer, Texture> skins;
    private final Region headRegion, hatRegion;
    
    public ChatRenderer(GameRenderer gameRenderer){
        this.gameRenderer = gameRenderer;
        this.textBatch = gameRenderer.getTextComponentBatch();
        this.batch = textBatch.getBatch();
        
        chatX = 10;
        chatY = 10;
        
        skins = new HashMap<>();
        for(int i = 1; i <= 21; i++)
            skins.put(i, new Texture("texture/skin/skin" + i + ".png"));
        
        headRegion = new Region(skins.get(1), 8, 8, 8, 8);
        hatRegion = new Region(skins.get(1), 40, 8, 8, 8);
    }
    
    
    public void render(){
        batch.begin();
        textBatch.setBackgroundColor(0, 0, 0, 0);
        
        final float chatHeight = Pize.getHeight() / 2F;
        final float chatWidth = Pize.getWidth() / 2F;
        final float lineAdvance = textBatch.getFont().getLineAdvanceScaled();
        
        final Chat chat = gameRenderer.getSession().getGame().getChat();
        final List<ChatMessage> messages = chat.getMessages();
        
        final float openedChatY = chatY + (chat.isOpened() ? lineAdvance + 10 : 0);
        float chatMessagesHeight = 0;
        for(ChatMessage message: messages)
            chatMessagesHeight += textBatch.getFont().getBounds(message.getComponents().toString(), chatWidth).y;
        
        // Enter
        if(chat.isOpened()){
            final String enteringText = chat.getEnteringText();
            final float lineWidth = textBatch.getFont().getLineWidth(enteringText);
            
            batch.drawQuad(0.4, chatX, chatY, Math.max(lineWidth, chatWidth), lineAdvance);
            
            final float cursorLineWidth = textBatch.getFont().getLineWidth(enteringText.substring(0, chat.getCursorX()));
            
            textBatch.drawComponent(new Component().text(enteringText), chatX, chatY);
            
            batch.drawQuad(1, 1, 1, 1, chatX + cursorLineWidth, chatY, textBatch.getFont().scale, lineAdvance);
        }
        
        // Scroll
        if(chat.isOpened() && Pize.mouse().isInBounds(chatX, openedChatY, chatWidth, chatHeight))
            scrollMotion -= Pize.mouse().getScroll() * Pize.getDt() * lineAdvance * 10;
        
        scroll += scrollMotion;
        scrollMotion *= 0.95;
        
        if(!chat.isOpened())
            scroll = 0;
        
        scroll = Math.min(0, scroll);
        scroll = Math.max(Math.min(0, chatHeight - chatMessagesHeight), scroll);
        
        // Chat
        final float headSize = lineAdvance;
        final float headAdvance = headSize * 1.5F;
        final float headDrawX = (headAdvance - headSize) / 2;
        final float headDrawY = (lineAdvance - headSize) / 2;
        
        batch.scissor.begin(0, chatX, openedChatY, chatWidth + headAdvance, chatHeight); // Scissors begin
        
        int textAdvanceY = 0;
        for(int i = messages.size() - 1; i >= 0; i--){
            final ChatMessage message = messages.get(i);
            
            float alpha = 1F;
            if(!chat.isOpened()){
                if(message.getSeconds() < MSG_LIFE_TIME_SEC)
                    alpha = (float) Math.min(1, MSG_LIFE_TIME_SEC - message.getSeconds());
                else
                    continue;
            }
            
            final float textHeight = textBatch.getFont().getTextHeight(message.getComponents().toString(), chatWidth + headAdvance);
            final float renderChatY = openedChatY + textAdvanceY + scroll;
            final float lineWrapAdvanceY = textHeight - lineAdvance;
            
            final boolean isPlayer = message.getSource().isPlayer();
            final StringBuilder messageStringBuilder = new StringBuilder();
            for(ComponentText component: message.getComponents())
                messageStringBuilder.append(component.toString());
            
            // Render background
            batch.drawQuad(0.4 * alpha, chatX, renderChatY, chatWidth + headAdvance, textHeight);
            
            // Render head
            if(isPlayer){
                final int skinID = Maths.abs(message.getSource().asPlayer().getPlayerName().hashCode()) % 20 + 1;
                final Texture skin = skins.get(skinID);
                
                batch.setAlpha(alpha);
                batch.draw(skin, chatX + headDrawX, renderChatY + lineWrapAdvanceY + headDrawY, headSize, headSize, headRegion);
                
                batch.setTransformOrigin(0.5, 0.5);
                batch.scale(1.1F);
                batch.draw(skin, chatX + headDrawX, renderChatY + lineWrapAdvanceY + headDrawY, headSize, headSize, hatRegion);
                batch.scale(1);
            }
            // Render text
            textBatch.drawComponents(message.getComponents(), chatX + headAdvance, renderChatY + lineWrapAdvanceY, chatWidth, alpha);
            
            textAdvanceY += textHeight;
        }
        
        batch.end();
        batch.scissor.end(0); // Scissors end
    }
    
    @Override
    public void dispose(){
        batch.dispose();
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/client/renderer/chat/ChatRenderer.kt
package pize.tests.voxelgame.client.renderer.chat

import pize.Pize.dt
import pize.Pize.height
import pize.Pize.mouse
import pize.Pize.width
import pize.app.Disposable
import pize.graphics.texture.Region
import pize.graphics.texture.Texture
import pize.graphics.util.batch.TextureBatch
import pize.math.Maths.abs
import pize.tests.voxelgame.client.renderer.GameRenderer
import pize.tests.voxelgame.client.renderer.text.TextComponentBatch
import pize.tests.voxelgame.main.text.Component
import kotlin.math.max
import kotlin.math.min

class ChatRenderer(private val gameRenderer: GameRenderer) : Disposable {
    private val batch: TextureBatch?
    private val textBatch: TextComponentBatch?
    private val chatX = 10
    private val chatY = 10
    private var scroll = 0f
    private var scrollMotion = 0f
    private val skins: MutableMap<Int, Texture>
    private val headRegion: Region
    private val hatRegion: Region

    init {
        textBatch = gameRenderer.textComponentBatch
        batch = textBatch.getBatch()
        skins = HashMap()
        for (i in 1..21) skins[i] = Texture("texture/skin/skin$i.png")
        headRegion = Region(skins[1]!!, 8.0, 8.0, 8.0, 8.0)
        hatRegion = Region(skins[1]!!, 40.0, 8.0, 8.0, 8.0)
    }

    fun render() {
        batch!!.begin()
        textBatch!!.setBackgroundColor(0.0, 0.0, 0.0, 0.0)
        val chatHeight = height / 2f
        val chatWidth = width / 2f
        val lineAdvance = textBatch.font.lineAdvanceScaled
        val chat = gameRenderer.session.game.chat
        val messages = chat.messages
        val openedChatY: Float = chatY + if (chat!!.isOpened) lineAdvance + 10 else 0
        var chatMessagesHeight = 0f
        for (message in messages!!) chatMessagesHeight += textBatch.font.getBounds(
            message.components.toString(),
            chatWidth.toDouble()
        ).y

        // Enter
        if (chat!!.isOpened) {
            val enteringText = chat.enteringText
            val lineWidth = textBatch.font.getLineWidth(enteringText)
            batch.drawQuad(
                0.4,
                chatX.toFloat(),
                chatY.toFloat(),
                max(lineWidth.toDouble(), chatWidth.toDouble()),
                lineAdvance
            )
            val cursorLineWidth = textBatch.font.getLineWidth(enteringText!!.substring(0, chat.cursorX))
            textBatch.drawComponent(
                Component().text(enteringText),
                chatX.toFloat(),
                chatY.toFloat()
            )
            batch.drawQuad(
                1.0,
                1.0,
                1.0,
                1.0,
                chatX + cursorLineWidth,
                chatY.toFloat(),
                textBatch.font.scale,
                lineAdvance
            )
        }

        // Scroll
        if (chat.isOpened && mouse()!!
                .isInBounds(chatX.toDouble(), openedChatY.toDouble(), chatWidth.toDouble(), chatHeight.toDouble())
        ) scrollMotion -= mouse()!!.scroll * dt * lineAdvance * 10
        scroll += scrollMotion
        scrollMotion *= 0.95.toFloat()
        if (!chat.isOpened) scroll = 0f
        scroll = min(0.0, scroll.toDouble()).toFloat()
        scroll = max(min(0.0, (chatHeight - chatMessagesHeight).toDouble()), scroll.toDouble())
            .toFloat()

        // Chat
        val headAdvance = lineAdvance * 1.5f
        val headDrawX = (headAdvance - lineAdvance) / 2
        val headDrawY = (lineAdvance - lineAdvance) / 2
        batch.scissor.begin(
            0,
            chatX.toDouble(),
            openedChatY.toDouble(),
            (chatWidth + headAdvance).toDouble(),
            chatHeight.toDouble()
        ) // Scissors begin
        var textAdvanceY = 0
        for (i in messages.indices.reversed()) {
            val message = messages[i]
            var alpha = 1f
            if (!chat.isOpened) {
                alpha = if (message.seconds < MSG_LIFE_TIME_SEC) min(
                    1.0,
                    MSG_LIFE_TIME_SEC - message.seconds
                )
                    .toFloat() else continue
            }
            val textHeight =
                textBatch.font.getTextHeight(message.components.toString(), (chatWidth + headAdvance).toDouble())
            val renderChatY = openedChatY + textAdvanceY + scroll
            val lineWrapAdvanceY = textHeight - lineAdvance
            val isPlayer = message.source.isPlayer
            val messageStringBuilder = StringBuilder()
            for (component in message.components) messageStringBuilder.append(component.toString())

            // Render background
            batch.drawQuad(0.4 * alpha, chatX.toFloat(), renderChatY, chatWidth + headAdvance, textHeight)

            // Render head
            if (isPlayer) {
                val skinID = abs(message.source.asPlayer().playerName.hashCode()) % 20 + 1
                val skin = skins[skinID]
                batch.setAlpha(alpha.toDouble())
                batch.draw(
                    skin,
                    chatX + headDrawX,
                    renderChatY + lineWrapAdvanceY + headDrawY,
                    lineAdvance,
                    lineAdvance,
                    headRegion
                )
                batch.setTransformOrigin(0.5, 0.5)
                batch.scale(1.1f)
                batch.draw(
                    skin,
                    chatX + headDrawX,
                    renderChatY + lineWrapAdvanceY + headDrawY,
                    lineAdvance,
                    lineAdvance,
                    hatRegion
                )
                batch.scale(1f)
            }
            // Render text
            textBatch.drawComponents(
                message.components,
                chatX + headAdvance,
                renderChatY + lineWrapAdvanceY,
                chatWidth,
                alpha
            )
            textAdvanceY = (textAdvanceY + textHeight).toInt()
        }
        batch.end()
        batch.scissor.end(0) // Scissors end
    }

    override fun dispose() {
        batch!!.dispose()
    }

    companion object {
        private const val MSG_LIFE_TIME_SEC = 6f
    }
}
